Add needs_exact_strides operator tag for Inductor to force exact strides

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #150511
* #148104
* #150495
* #148092
* #148091
* __->__ #148063
* #148046

Inductor will force exact strides on a custom operator tagged with
needs_exact_strides. I'll make this the default in a follow-up PR.

Test Plan:
- tests

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

zou3519 committed:
Add needs_exact_strides operator tag for Inductor to force exact strides

Inductor will force exact strides on a custom operator tagged with
needs_exact_strides. I'll make this the default in a follow-up PR.

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add needs_exact_strides operator tag for Inductor to force exact strides"


Inductor will force exact strides on a custom operator tagged with
needs_exact_strides. I'll make this the default in a follow-up PR.

Test Plan:
- tests

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

zou3519 committed:
Update on "Add needs_exact_strides operator tag for Inductor to force exact strides"


Inductor will force exact strides on a custom operator tagged with
needs_exact_strides. I'll make this the default in a follow-up PR.

Test Plan:
- tests

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

zou3519 committed:
Update on "Add needs_exact_strides operator tag for Inductor to force exact strides"


Inductor will force exact strides on a custom operator tagged with
needs_exact_strides. I'll make this the default in a follow-up PR.

Test Plan:
- tests

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

zou3519 committed:
Update on "Add needs_exact_strides operator tag for Inductor to force exact strides"


Inductor will force exact strides on a custom operator tagged with
needs_exact_strides. I'll make this the default in a follow-up PR.

Test Plan:
- tests

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

zou3519 committed:
Update on "Add needs_exact_strides operator tag for Inductor to force exact strides"


Inductor will force exact strides on a custom operator tagged with
needs_exact_strides. I'll make this the default in a follow-up PR.

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add needs_exact_strides operator tag for Inductor to force exact strides"


Inductor will force exact strides on a custom operator tagged with
needs_exact_strides. I'll make this the default in a follow-up PR.

Test Plan:
- tests

[ghstack-poisoned]

etaf commented on test/inductor/test_triton_kernels.py:
```diff
@@ -3442,12 +3441,48 @@ def decomp(*flat_args):
+            def impl2(x):
+                return x.clone(memory_format=torch.channels_last)
+
+            lib.impl("force_channels_last", impl2, "CUDA")
```
Hi, I found here is hard code "CUDA", this would fail on XPU, can we mark this test case as requires CUDA or use `GPU_TYPE.upper()` to generalize the hard code ?

zou3519 committed:
Update on "Add needs_exact_strides operator tag for Inductor to force exact strides"


Inductor will force exact strides on a custom operator tagged with
needs_exact_strides. I'll make this the default in a follow-up PR.

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add needs_exact_strides operator tag for Inductor to force exact strides"


Inductor will force exact strides on a custom operator tagged with
needs_exact_strides. I'll make this the default in a follow-up PR.

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add needs_exact_strides operator tag for Inductor to force exact strides"


Inductor will force exact strides on a custom operator tagged with
needs_exact_strides. I'll make this the default in a follow-up PR.

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add needs_exact_strides operator tag for Inductor to force exact strides"


Inductor will force exact strides on a custom operator tagged with
needs_exact_strides. I'll make this the default in a follow-up PR.

Test Plan:
- tests

[ghstack-poisoned]
